Add CA1868: Unnecessary call to `Set.Contains(item)`

mpidash opened the pull request:
This analyzer detects when either `Set.Add` or `Set.Remove` is guarded by `Set.Contains`.
As recommended in dotnet/runtime#85490, a fix is only suggested when the conditional body only has a single `Set.Add` or `Set.Remove`.

I tried out the analyzer in `dotnet/roslyn`, `dotnet/roslyn-analyzers` and `dotnet/runtime` and found cases of guarded calls:

1. [roslyn-analyzers/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#1376](https://github.com/dotnet/roslyn-analyzers/blob/40a3ec17a7507b8767b4ac806f29b6df179cf627/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#L1376)
2. [runtime/src/tasks/MobileBuildTasks/Android/AndroidProject.cs#L125](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/tasks/MobileBuildTasks/Android/AndroidProject.cs#L125)
3. [runtime/src/tools/illink/src/ILLink.Shared/Annotations.cs#L82](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/tools/illink/src/ILLink.Shared/Annotations.cs#L82)
4. [runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleInitializerListNode.cs#L116](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleInitializerListNode.cs#L116)
5. [runtime/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs#L475](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs#L475)
6. [runtime/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DgmlWriter.cs#L161](https://github.com/dotnet/runtime/blob/ddbce91b241bcf0d7f33d04d33520659addd6dd7/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DgmlWriter.cs#L161)
7. [roslyn/src/Compilers/Core/CodeAnalysisTest/Collections/HashSet/ISet_Generic_Tests`1.cs#L595](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/Core/CodeAnalysisTest/Collections/HashSet/ISet_Generic_Tests%601.cs#L595)
8. [roslyn/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2073](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2073)
9. [roslyn/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2094](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2094)
10. [roslyn/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb#L1805](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb#L1805)
11. ~[roslyn/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb#L958](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb#L958)~ No longer flagged as it uses a property which could cause side effects.
12. [roslyn/src/EditorFeatures/Core/CodeDefinitionWindow/DefinitionContextTracker.cs#L80](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/EditorFeatures/Core/CodeDefinitionWindow/DefinitionContextTracker.cs#L80)
13. [roslyn/src/VisualStudio/Core/Def/ChangeSignature/ChangeSignatureDialogViewModel.cs#L244](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/VisualStudio/Core/Def/ChangeSignature/ChangeSignatureDialogViewModel.cs#L244)
14. [roslyn/src/VisualStudio/Core/Def/Interop/CleanableWeakComHandleTable.cs#L173](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/VisualStudio/Core/Def/Interop/CleanableWeakComHandleTable.cs#L173)
15. [roslyn/src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/VisualBasicProject.vb#L341](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/VisualBasicProject.vb#L341)

Fixes dotnet/runtime#85490

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpDoNotGuardSetAddOrRemoveByContains.Fixer.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace Microsoft.NetCore.CSharp.Analyzers.Performance
+{
+    internal class CSharpDoNotGuardSetAddOrRemoveByContainsFixer : DoNotGuardSetAddOrRemoveByContainsFixer
```
This is missing ExportCodeFixProvider and Shared attributes.

Also consider making the class `sealed`

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs:
```diff
@@ -0,0 +1,200 @@
+                    return false;
+            }
+
+            return DoesSignatureMatch(containsInvocation.TargetMethod, containsMethod);
```
Why not compare the symbols directly?

mpidash replied:
The `containsMethod` comes from `ICollection<T>` whereas `containsInvocation.TargetMethod` comes from the concrete implementation (e.g. `HashSet<T>`).
So comparing the symbols directly would lead to inequality, as `System.Collections.Generic.ICollection<T>.Contains(T)` != `System.Collections.Generic.HashSet<string>.Contains(string)`.

That is why I resorted to compare method signatures instead.

But I think a better approach would be to check if the target method implements the interface method, something like this:
```C#
private static bool DoesImplementInterfaceMethod(IMethodSymbol method, IMethodSymbol interfaceMethod)
{
    return method.IsImplementationOfInterfaceMethod(method.Parameters[0].Type, interfaceMethod.ContainingType, interfaceMethod.Name);
}
```

`PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer` also uses a signature check which could be replaced by the above method:

https://github.com/dotnet/roslyn-analyzers/blob/4592065e23f7c84047b4e798ec41efd9876c41bc/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs#L263-L268

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.Fixer.cs:
```diff
@@ -0,0 +1,63 @@
+                return;
+            }
+
+            var diagnostic = context.Diagnostics.First();
```
```suggestion
            var diagnostic = context.Diagnostics[0];
```

mpidash replied:
Done

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.Fixer.cs:
```diff
@@ -0,0 +1,63 @@
+            var title = MicrosoftNetCoreAnalyzersResources.DoNotGuardSetAddOrRemoveByContainsTitle;
+            var codeAction = CodeAction.Create(title,
+                ct => Task.FromResult(ReplaceConditionWithChild(context.Document, root, conditionalSyntax, childStatementSyntax)),
+                title);
```
```suggestion
                nameof(MicrosoftNetCoreAnalyzersResources.DoNotGuardSetAddOrRemoveByContainsTitle));
```

mpidash replied:
Done

mpidash commented:
I've added additional tests and corrected the fixer analogous to #6770, PTAL.

buyaa-n commented:
> I tried out the analyzer in dotnet/roslyn, dotnet/roslyn-analyzers and dotnet/runtime and found cases of guarded calls:
> 
> [roslyn-analyzers/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#1376](https://github.com/dotnet/roslyn-analyzers/blob/40a3ec17a7507b8767b4ac806f29b6df179cf627/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#L1376)
> [runtime/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DgmlWriter.cs#L161](https://github.com/dotnet/runtime/blob/ddbce91b241bcf0d7f33d04d33520659addd6dd7/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DgmlWriter.cs#L161)
> ...

Thank you @mpidash I found more hits in runtime, FYI you could build with `-warnaserror 0` option to continue building with analyzers warnings, to see all the diagnostics caused by the analyzer. 

Found one false positive with the testing, it warns when the elements in the `Contains` and `Add` area different, for example:

```cs
    public void Test(HashSet<string> runtimeOptions, string[] jiterpreterOptions, bool jiterpreter)
    {
        foreach (var jiterpreterOption in jiterpreterOptions)
        {
            if (jiterpreter == true)
            {
                if (!runtimeOptions.Contains($"--no-{jiterpreterOption}"))
                    runtimeOptions.Add($"--{jiterpreterOption}");
            }
            else
            {
                if (!runtimeOptions.Contains($"--{jiterpreterOption}"))
                    runtimeOptions.Add($"--no-{jiterpreterOption}");
            }
        }
    }
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1509,6 +1509,15 @@
+    <value>Do not guard 'Set.Add(item)' or 'Set.Remove(item)' with 'Set.Contains(item)'</value>
+  </data>
+  <data name="DoNotGuardSetAddOrRemoveByContainsTitle" xml:space="preserve">
+    <value>Unnecessary call to 'Set.Contains(item)'</value>
```
Showing same title in the fixer do not look good, it should be something that illustrates what the fixer does, like the one below. It looks, you can even use the below `RemoveRedundantGuardCallCodeFixTitle` as is for the fixer

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1509,6 +1509,15 @@
     <value>Unnecessary call to 'Dictionary.ContainsKey(key)'</value>
   </data>
+  <data name="DoNotGuardSetAddOrRemoveByContainsDescription" xml:space="preserve">
+    <value>Do not guard 'Set.Add(item)' or 'Set.Remove(item)' with 'Set.Contains(item)'. The former two already check whether the item exists and will return if it was added or removed.</value>
```
Using `'Set.Add(item)', 'Set.Remove(item)', 'Set.Contains(item)'` in the messaging here and below doesn't look right. I understand the issue written like this to illustrate the Sets in general, but the Sets flagged here are not derived from `Set`, they derive from <ISet> which doesn't have `Remove/Contains` methods, so it might cause a confusion

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1509,6 +1509,15 @@
+    <value>Do not guard 'Set.Add(item)' or 'Set.Remove(item)' with 'Set.Contains(item)'. The former two already check whether the item exists and will return if it was added or removed.</value>
+  </data>
+  <data name="DoNotGuardSetAddOrRemoveByContainsMessage" xml:space="preserve">
+    <value>Do not guard 'Set.Add(item)' or 'Set.Remove(item)' with 'Set.Contains(item)'</value>
```
For messages we can pass the API names to the message, when creating the diagnostics
```suggestion
    <value>Do not guard '{0}' with '{1}'</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs:
```diff
@@ -0,0 +1,197 @@
+            }
+
+            addMethod = iSetType.GetMembers(Add).OfType<IMethodSymbol>().FirstOrDefault();
+            containsMethod = iCollectionType.GetMembers(Contains).OfType<IMethodSymbol>().FirstOrDefault();
+            removeMethod = iCollectionType.GetMembers(Remove).OfType<IMethodSymbol>().FirstOrDefault();
```
Could use just one loop
```suggestion
            foreach (var method in iCollectionType.GetMembers().OfType<IMethodSymbol>())
            {
                switch (method.Name)
                {
                    case Remove: removeMethod = method; break;
                    case Contains: containsMethod = method; break;
                }
            }
```

mpidash replied:
The analyzer needs both `Remove` and `Contains` from `ICollection<T>` -- this would only find one of them and then break out.

buyaa-n replied:
No, the break is for switch case not for the loop

mpidash replied:
Ohh, I see. I missed that, sorry. I will also apply the same change for `iImmutableSetType` and its methods.

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContainsTests.cs:
```diff
@@ -0,0 +1,841 @@
+            {
+                MySet.Add(""Item"");
+                Console.WriteLine();
+            }
```
Simple additional statements like this could have code fix suggestions, but it is not required for this PR, we can file an issue for further improvement

mpidash commented:
> Thank you @mpidash I found more hits in runtime, FYI you could build with `-warnaserror 0` option to continue building with analyzers warnings, to see all the diagnostics caused by the analyzer.

Thanks, that will be very useful for testing 😄 

> Found one false positive with the testing, it warns when the elements in the `Contains` and `Add` area different, for example:

I addressed different arguments in [d7c7c54](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/d7c7c54e1c1ee25170fe606efbe2cab11d83209f).
The analyzer for dictionaries (`DoNotGuardDictionaryRemoveByContainsKey`) has the same problem, I will create an issue.

mpidash commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs:
```diff
@@ -0,0 +1,197 @@
+            [NotNullWhen(true)] out IMethodSymbol? removeMethod)
+        {
+            var iSetType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericISet1);
+            var iCollectionType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericICollection1);
```
`ImmutableHashSet.Builder` and `ImmutableSortedSet.Builder` work (added a test in cf49578).
I also added a test for the expected behavior for `Add` and `Delete` from `IImmutableSet<T>` (only reporting a diagnostic).

But I couldn't get it to work yet, the WIP commit is 4f62beb.

For some reason the following fails when `method` is `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)` and `interfaceMethod` is `System.Collections.Immutable.IImmutableSet<T>.Add(T)`:

```C#
private static bool DoesImplementInterfaceMethod(IMethodSymbol method, IMethodSymbol interfaceMethod)
{
    return method.IsImplementationOfInterfaceMethod(method.Parameters[0].Type, interfaceMethod.ContainingType, interfaceMethod.Name);
}
```

I stepped into `IsImplementationOfInterfaceMethod` which lead to this:
![image](https://github.com/dotnet/roslyn-analyzers/assets/4247811/08a57975-4f15-47b1-90be-a951857ade87)

I don't know why the symbol returned by `FindImplementationForInterfaceMember` is `System.Collections.Immutable.ImmutableHashSet<string>.System.Collections.Immutable.IImmutableSet<string>.Add(string)`. I would except `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)`. This will result in a false comparison result.

buyaa-n replied:
> ImmutableHashSet.Builder and ImmutableSortedSet.Builder work (added a test in https://github.com/dotnet/roslyn-analyzers/commit/cf4957820e06c8f440c382ac015c590a1b10abfd).
> I also added a test for the expected behavior for Add and Delete from IImmutableSet<T> (only reporting a diagnostic).

Thank you!

> For some reason the following fails when method is `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)` and interfaceMethod is `System.Collections.Immutable.IImmutableSet<T>.Add(T)`:

Because `System.Collections.Immutable.ImmutableHashSet<string>.Add(string)` is not implemented `System.Collections.Immutable.IImmutableSet<T>.Add(T)` instead it has explicit implementation of it instead https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablehashset-1.system-collections-immutable-iimmutableset-t--add?view=net-7.0#system-collections-immutable-immutablehashset-1-system-collections-immutable-iimmutableset((t))-add(-0)

mpidash replied:
I replaced the call to check interface implementations with checking the original definitions (which also contain explicit interface implementations). All tests, including the ones checking for `ImmutableHashSet` and `ImmutableSortedSet`, pass with the following change:

```C#
private static bool DoesImplementInterfaceMethod(IMethodSymbol method, IMethodSymbol interfaceMethod)
{
    var typedInterface = interfaceMethod.ContainingType.Construct(method.Parameters[0].Type);
    var typedInterfaceMethod = typedInterface.GetMembers(interfaceMethod.Name).FirstOrDefault();

    // Check against all original definitions to also cover external interface implementations
    return method.GetOriginalDefinitions().Any(definition => SymbolEqualityComparer.Default.Equals(definition, typedInterfaceMethod));
}
```

buyaa-n commented:
> The analyzer for dictionaries (`DoNotGuardDictionaryRemoveByContainsKey`) has the same problem ~I will create an issue~ => #6781.

Thanks for checking that analyzer and filing the issue

mpidash commented:
I re-tested the analyser against `dotnet/roslyn`, `dotnet/roslyn-analyzers` and `dotnet/runtime`, and found a couple of bugs that I fixed in the previous commits:

1. `DoesImplementInterfaceMethod` was not checking the parameter length of the method, resulting in an exception ([f22e2dc](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/f22e2dc55fe7b4b7b596c3fb1e815c5ae41a34bb))
2. The change to check all descendants caused nested conditionals to be flagged. I have reverted and added a case for `ISimpleAssignmentOperation` and `IVariableDeclarationGroupOperation` ([efe2993](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/efe299362a536940b5078246e046941fb741a891))
4. I forgot to add the format arguments for the message ([7c68257](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/7c68257b5c822c7bb5722f3dba65b761b80c5767))
5. No diagnostic was created when accessing the set through the interface types ([860f217](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/860f21706663fce70aa358e50254b302b7e1990a))
6. `IImmutableSet<T>.Contains` was not detected. I have changed the analyzer to use a helper class `RequiredSymbols`, as seen in others (sorry for the big change so late in the PR). This helper now also gets `IImmutableSet<T>.Contains` and uses it for checks ([8b7a007](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/8b7a007e5076c57c75d8edb43e5e27fd826fd48d)).

I also updated the findings:

> 1. [roslyn-analyzers/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#1376](https://github.com/dotnet/roslyn-analyzers/blob/40a3ec17a7507b8767b4ac806f29b6df179cf627/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#L1376)
> 2. [runtime/src/tasks/MobileBuildTasks/Android/AndroidProject.cs#L125](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/tasks/MobileBuildTasks/Android/AndroidProject.cs#L125)
> 3. [runtime/src/tools/illink/src/ILLink.Shared/Annotations.cs#L82](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/tools/illink/src/ILLink.Shared/Annotations.cs#L82)
> 4. [runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleInitializerListNode.cs#L116](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleInitializerListNode.cs#L116)
> 5. [runtime/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs#L475](https://github.com/dotnet/runtime/blob/a2ce8773521a111154e30ab0abe292d9f38d13e6/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs#L475)
> 6. [runtime/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DgmlWriter.cs#L161](https://github.com/dotnet/runtime/blob/ddbce91b241bcf0d7f33d04d33520659addd6dd7/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DgmlWriter.cs#L161)
> 7. [roslyn/src/Compilers/Core/CodeAnalysisTest/Collections/HashSet/ISet_Generic_Tests`1.cs#L595](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/Core/CodeAnalysisTest/Collections/HashSet/ISet_Generic_Tests%601.cs#L595)
> 8. [roslyn/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2073](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2073)
> 9. [roslyn/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2094](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs#L2094)
> 10. [roslyn/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb#L1805](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb#L1805)
> 11. ~[roslyn/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb#L958](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb#L958)~ No longer flagged as it uses a property which could cause side effects.
> 12. [roslyn/src/EditorFeatures/Core/CodeDefinitionWindow/DefinitionContextTracker.cs#L80](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/EditorFeatures/Core/CodeDefinitionWindow/DefinitionContextTracker.cs#L80)
> 13. [roslyn/src/VisualStudio/Core/Def/ChangeSignature/ChangeSignatureDialogViewModel.cs#L244](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/VisualStudio/Core/Def/ChangeSignature/ChangeSignatureDialogViewModel.cs#L244)
> 14. [roslyn/src/VisualStudio/Core/Def/Interop/CleanableWeakComHandleTable.cs#L173](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/VisualStudio/Core/Def/Interop/CleanableWeakComHandleTable.cs#L173)
> 15. [roslyn/src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/VisualBasicProject.vb#L341](https://github.com/dotnet/roslyn/blob/31dff86c598c52f4cf4eccfe2c6be713de9c7be0/src/VisualStudio/VisualBasic/Impl/ProjectSystemShim/VisualBasicProject.vb#L341)

`11.` is no longer found because it uses a property as item, `2.` to `5.` are new. I found no false positives.
The message now looks like this:

```log
W:\runtime\src\coreclr\tools\aot\ILCompiler.DependencyAnalysisFramework\DgmlWriter.cs(161,18): warning CA1865: Do not guard 'HashSet.Add(Tuple)' with 'HashSet.Contains(Tuple)'
W:\roslyn\src\VisualStudio\VisualBasic\Impl\ProjectSystemShim\VisualBasicProject.vb(341,28): info CA1865: Do not guard 'HashSet.Remove(String)' with 'HashSet.Contains(String)'
```

jeffhandley reviewed (approved):
I'm supportive of this merging in for .NET 8 RC1.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1509,6 +1509,15 @@
+    <value>Do not guard '{0}' with '{1}'</value>
+  </data>
+  <data name="DoNotGuardSetAddOrRemoveByContainsTitle" xml:space="preserve">
+    <value>Remove unnecessary call</value>
```
@mpidash it seems you misunderstood my [comment](https://github.com/dotnet/roslyn-analyzers/pull/6767#discussion_r1267102149) `Showing same title in the fixer do not look good, it should be something that illustrates what the fixer does, like the one below. It looks, you can even use the below RemoveRedundantGuardCallCodeFixTitle as is for the fixer` 

I meant to use different title for analyzer and code fixer. The analyzer title can be kept as it was `Unnecessary call to 'Set.Contains(item)'` or omit Set: `Unnecessary call to 'Contains(item)'`. Then the code fixer should be illustrate the action it is fixing i.e.: `Remove unnecessary call` for that don't need to create another title with same value, can just use the `RemoveRedundantGuardCallCodeFixTitle` below

mpidash replied:
Oh, sorry about that and thanks for the clarification. I have dropped the `Set` in the message as you suggested to be consistent with `DoNotGuardSetAddOrRemoveByContainsDescription`.

Changed in 5fef8d0.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs:
```diff
@@ -0,0 +1,347 @@
+                locations.Add(addOrRemoveInvocation.Syntax.Parent!.GetLocation());
+
+                // Build custom format instead of CSharpShortErrorMessageFormat/VisualBasicShortErrorMessageFormat to prevent unhelpful messages for VB.
+                var symbolDisplayFormat = SymbolDisplayFormat.MinimallyQualifiedFormat
+                    .WithParameterOptions(SymbolDisplayParameterOptions.IncludeType)
+                    .WithGenericsOptions(SymbolDisplayGenericsOptions.None)
+                    .WithMemberOptions(SymbolDisplayMemberOptions.IncludeParameters | SymbolDisplayMemberOptions.IncludeContainingType)
+                    .WithKindOptions(SymbolDisplayKindOptions.None);
```
Doesn't seem need to create for each conditional, can be moved to static readonly field

mpidash replied:
Changed in [8a3f6ef](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/8a3f6ef034caf207bb31272a511b58d877985744).

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs:
```diff
@@ -0,0 +1,347 @@
+        //   2. Identical: Contains(item) == Add(item), Contains("const") == Add("const")
+        private static bool AreInvocationArgumentsEqual(IInvocationOperation invocation1, IInvocationOperation invocation2)
+        {
+            return invocation1.Arguments
+                .Zip(invocation2.Arguments, (a1, a2) => IsArgumentValueEqual(a1.Value, a2.Value))
+                .All(argumentsEqual => argumentsEqual);
```
Can't we just compare the 1st arguments? I think all involved methods has only 1 argument
```suggestion
            return IsArgumentValueEqual(invocation1.Arguments[0].Value, invocation2.Arguments[0].Value);
```

buyaa-n commented:
> I re-tested the analyser against dotnet/roslyn, dotnet/roslyn-analyzers and dotnet/runtime, and found a couple of bugs that I fixed in the previous commits:

Great thank you @mpidash feel free to put up a PR with the fix for valid diagnostics to those repos

> The message now looks like this:
 
Looks great, thank you!

mpidash commented:
> Great thank you @mpidash feel free to put up a PR with the fix for valid diagnostics to those repos

I'll prepare one for `dotnet/runtime`, already submitted for the other two: https://github.com/dotnet/roslyn/pull/69179 and #6793.
 
> Looks great, thank you! Left a few more comment, overall the PR shape looks good.

Thank you so much for your guidance and patience with me! 🎉

sharwell commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpDoNotGuardSetAddOrRemoveByContains.Fixer.cs:
```diff
@@ -0,0 +1,79 @@
+
+                newRoot = root.ReplaceNode(conditionalOperationNode, newConditionalOperationNode);
+            }
+            else if (conditionalOperationNode is IfStatementSyntax ifStatementSyntax && ifStatementSyntax.Else != null)
```
```suggestion
            else if (conditionalOperationNode is IfStatementSyntax { Else: not null } ifStatementSyntax)
```

sharwell commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpDoNotGuardSetAddOrRemoveByContains.Fixer.cs:
```diff
@@ -0,0 +1,79 @@
+                SyntaxNode newConditionalOperationNode = conditionalExpressionSyntax
+                    .WithCondition((ExpressionSyntax)expression)
+                    .WithWhenTrue(conditionalExpressionSyntax.WhenFalse)
+                    .WithWhenFalse(null!)
```
❗ This is going to throw ArgumentNullException

buyaa-n replied:
Good catch, I thought this is for removing the false part, did not notice the '!'. What can we use instead?

mpidash replied:
I originally copied this section from the similar fixer for dictionaries (`CSharpDoNotGuardDictionaryRemoveByContainsKeyFixer`).
As part of a bugfix for the dictionary analyzer, it was discovered that the case of the ternary operator was never flagged by the analyzer (see https://github.com/dotnet/roslyn-analyzers/pull/6770#discussion_r1264964323).
So the code in the fixer that would throw an ArgumentNullException is never executed.

That part should probably be removed from the dictionary fixer as well (I missed that in the PR above):
https://github.com/dotnet/roslyn-analyzers/blob/47edf21cd38fb583da54085bdfd724c2cdf1f6f0/src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpDoNotGuardDictionaryRemoveByContainsKeyFixer.cs#L44-L48

buyaa-n replied:
Thanks for explanation @mpidash, removing this part sounds good to me, what you think @sharwell?

sharwell replied:
I believe if this is reachable, it would would be hit by:

```
bool added = set.Contains(value) ? false : set.Add(value);
```

mpidash replied:
I've added two tests for ternary operators:

1. One with a ternary operator as in your example. In this case, no diagnostic is reported.
2. Another test with a nested ternary operator. No diagnostic is reported for the `Add` case, but one is reported for the `Remove` case.

This is because we only check the true part of a conditional for an applicable `Add` or `Remove`. So the `Add` case gets filtered out immediately because it contains no child operations:

```C#
var firstOperation = operations.FirstOrDefault();

if (firstOperation is null)
{
    return false;
}
```

For `Remove` it depends on the content of the true part. In the simple case, it contains a `FieldReferenceOperation`, which is not covered by the switch in `HasApplicableAddOrRemoveMethod` so it produces no diagnostic.
The nested case for `Remove` produces a diagnostic because the first child operation is a `IInvocationOperation`.

A ternary operator gets filtered out by the fixer in `SyntaxSupportedByFixer`, because the `childStatementSyntax` is not a `ExpressionStatementSyntax` (it is `ConditionalExpressionSyntax`). So the part where the `ArgumentNullException` would be thrown is not reachable.

I think we should filter out ternary operators for now. I am not sure how likely it is that someone would write code like in the test cases, as they already use the return value of `Add` and `Remove`.
`DoNotGuardDictionaryRemoveByContainsKey` should be adapted to treat ternary operators in the same way as this analyzer.

---

The above has shown a case that is not covered at the moment:

```C#
if (MySet.Contains(""Item""))
{
    throw new Exception(""Item already exists"");
}
else
{
    MySet.Add(""Item"");
}
```

which could be refactored to:

```C#
if (!MySet.Add(""Item""))
{
    throw new Exception(""Item already exists"");
}
```

buyaa-n replied:
> I think we should filter out ternary operators for now. I am not sure how likely it is that someone would write code like in the test cases, as they already use the return value of Add and Remove.
> DoNotGuardDictionaryRemoveByContainsKey should be adapted to treat ternary operators in the same way as this analyzer.

Agree that it would be rare, though it could happen and it fixable. I am OK with filtering this out for now and create an issue for future fix.

> The above has shown a case that is not covered at the moment:

Did you find why it was not covered and could you fix this with the PR?

mpidash replied:
> Did you find why it was not covered and could you fix this with the PR?

Its because atm only the true branch of a condition is checked for `Add` and `Remove`.
I will add a fix which also checks the false branch.

mpidash replied:
- [ad82982](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/ad8298285efd463cd5ac557387ff2ddc35267225) adds support for `Add` and `Remove` in else blocks
- [f723ac5](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/f723ac580879390b3ef4762bee690c4594555caf) adds support for ternary diagnostics (no fixer)

buyaa-n replied:
Thank you for the fix @mpidash!

@sharwell requested change is addressed PTAL

buyaa-n commented:
@mpidash seems need to run `msbuild -pack` once more

sharwell commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContainsTests.cs:
```diff
@@ -0,0 +1,1159 @@
+        }
+        #endregion
+
+        #region Helpers
+        private const string CSUsings = @"using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;";
+
+        private const string CSNamespaceAndClassStart = @"namespace Testopolis
+{
+    public class MyClass
+    {
+";
+
+        private const string CSNamespaceAndClassEnd = @"
+    }
+}";
+
+        private const string VBUsings = @"Imports System
+Imports System.Collections.Generic
+Imports System.Collections.Immutable";
```
💡 It would be preferable to inline all of these

mpidash replied:
Just to be sure that I understand correctly:

Change this:

```C#
[Fact]
public async Task AddWithVariableAssignment_ReportsDiagnostic_CS()
{
    string source = CSUsings + CSNamespaceAndClassStart + @"
private readonly HashSet<string> MySet = new HashSet<string>();

public MyClass()
{
    if (![|MySet.Contains(""Item"")|])
    {
        bool result = MySet.Add(""Item"");
    }
}
" + CSNamespaceAndClassEnd;

    await VerifyCS.VerifyCodeFixAsync(source, source);
}
```

To:

```C#
[Fact]
public async Task AddWithVariableAssignment_ReportsDiagnostic_CS()
{
    string source = """
        using System.Collections.Generic;

        class C
        {
            private readonly HashSet<string> MySet = new HashSet<string>();

            void M()
            {
                if (![|MySet.Contains("Item")|])
                {
                    bool result = MySet.Add("Item");
                }
            }
        }
        """;

    await VerifyCS.VerifyCodeFixAsync(source, source);
}
```

buyaa-n replied:
> Just to be sure that I understand correctly:

Correct, it meant to move `CSNamespaceAndClassStart` and `CSNamespaceAndClassEnd` inline.

mpidash replied:
Changed in [6f69703](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/6f6970341a2f77e921f92ec32e6d050fa7bf780c).

sharwell commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContainsTests.cs:
```diff
@@ -0,0 +1,1159 @@
+        [Fact]
+        public async Task ContainsAndAddCalledWithSameArgumentsLocals_OffersFixer_CS()
+        {
+            string source = CSUsings + CSNamespaceAndClassStart + @"
+        private readonly HashSet<string> MySet = new HashSet<string>();
+
+        public MyClass()
+        {
+            const string LocalItem = ""Item"";
+
+            if (![|MySet.Contains(LocalItem)|])
+            {
+                MySet.Add(LocalItem);
+            }
+        }" + CSNamespaceAndClassEnd;
```
💡 It would be preferable to use raw strings instead of verbatim strings for all of these. If you only convert a subset, I would suggest starting with cases that are not parameterized.

mpidash replied:
Changed in [6f69703](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/6f6970341a2f77e921f92ec32e6d050fa7bf780c).

mpidash commented:
I've re-ran the analyzer against `dotnet/roslyn`, `dotnet/roslyn-analyzers` and `dotnet/runtime` and found two new guarded calls, both in else blocks:

1. [roslyn/src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs#L333-L340](https://github.com/dotnet/roslyn/blob/c4ffabab7de0c314cdf86ead9f86543fa39a0758/src/Compilers/CSharp/Portable/Parser/DocumentationCommentParser.cs#L333-L340)
2. [runtime/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePartDefinition.cs#L188-L195](https://github.com/dotnet/runtime/blob/08acbb7ad7154e9d18227e4b00e4c9f41af1fffa/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePartDefinition.cs#L188-L195)

Also ran it against `dotnet/aspnetcore` and found one guarded call: [aspnetcore/src/Shared/StaticWebAssets/ManifestStaticWebAssetFileProvider.cs#L103-L110](https://github.com/dotnet/aspnetcore/blob/ea3ce10efd605856914cfdc8ec3c50d559776116/src/Shared/StaticWebAssets/ManifestStaticWebAssetFileProvider.cs#L103-L110)

I found no cases of ternary operators.

mpidash commented:
Looks like CA1865 was used by another analyzer that was already merged -- I will switch to the next one that is free.

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContainsTests.cs:
```diff
@@ -0,0 +1,2367 @@
+        }
+
+        [Fact]
+        public async Task RemoveInTernaryWhenFalseWithNonNegatedContains_ReportsDiagnostic_CS()
```
```suggestion
        public async Task RemoveInTernaryWhenFalseWithNonNegatedContains_NoDiagnostic_CS()
```

buyaa-n reviewed (approved):
Thank you @mpidash LGTM!
